components/tikv_util/src/sequence_number.rs

```diff
@@ -0,0 +1,155 @@
+// Copyright 2022 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::{
+    cmp,
+    collections::{BTreeMap, VecDeque},
+    sync::atomic::{AtomicU64, Ordering},
+};
+
+use lazy_static::lazy_static;
+
+lazy_static! {
+    static ref SEQUENCE_NUMBER_COUNTER_ALLOCATOR: AtomicU64 = AtomicU64::new(0);
+    pub static ref VERSION_COUNTER_ALLOCATOR: AtomicU64 = AtomicU64::new(0);
+}
+
+pub trait Notifier: Send {
+    fn notify_seqno_version_updated(&self, version: u64);
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub struct SequenceNumber {
+    pub number: u64,
+    pub version: u64,
+    start_counter: u64,
+    end_counter: u64,
+}
+
+impl SequenceNumber {
+    pub fn start() -> Self {
+        SequenceNumber {
+            number: 0,
+            start_counter: SEQUENCE_NUMBER_COUNTER_ALLOCATOR.fetch_add(1, Ordering::SeqCst) + 1,
+            end_counter: 0,
+            version: 0,
+        }
+    }
+
+    pub fn end(&mut self, number: u64) {
+        self.number = number;
+        self.end_counter = SEQUENCE_NUMBER_COUNTER_ALLOCATOR.load(Ordering::SeqCst);
+        self.version = VERSION_COUNTER_ALLOCATOR.load(Ordering::SeqCst);
+    }
+
+    pub fn max(left: Self, right: Self) -> Self {
+        match left.number.cmp(&right.number) {
+            cmp::Ordering::Less | cmp::Ordering::Equal => right,
+            cmp::Ordering::Greater => left,
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct SequenceNumberWindow {
+    // The sequence number doesn't be received in order, we need a ordered set to
+    // store received start counters which are bigger than last_start_counter + 1.
+    pending_start_counter: VecDeque<bool>,
+    // counter start from 1, so 0 means no start counter received.
+    last_ack_counter: u64,
+    // (end_counter, sequence number)
+    pending_sequence: BTreeMap<u64, SequenceNumber>,
```

Comment:
can it be simplified with Vec instead. Because we can use end_counter - last_ack_counter -1 as index

Reply:
Is it doable?